### PR TITLE
SQLite structure and routines updated for orphaned blocks

### PR DESCRIPTION
**Pull request by buck54321**
For #747
The solution got a little bigger than I expected. Let me know if you want anything changed. The database went from 250 to 430 MB, which is more than I would have expected for essentially adding 3 columns in total. Maybe there's some optimization I'm missing. 

I did not duplicate the `is_mainchain` and `is_valid` columns on the stake table, but that means that a few queries required some `JOIN`s with the block summary table.

### CODE REVIEW

**Comment by chappjc** [comment]
@buck54321 Instead of doing a merge with master (or resolving conflicts with github's webpage, which causes such a merge) please next time rebase and resolve the conflicts on your local machine.  Thanks for the PR!

**Comment by buck54321** [comment]
@chappjc No problem. I'll get it right next time. Still figuring out this workflow.

**Comment by chappjc** [comment]
The CONTRIBUTING doc needs lots of work.  It neglects a lot of important details.

**Comment by buck54321** [comment]
I see issue #640. I'm mostly confused about the exact sequence for squashing. Intuition tells me to 

checkout branch x -> do stuff -> commit  -> pull from up-to-date master -> push to branch x on remote -> pull request

and I'm not certain where exactly certain where the squash/rebase is performed.  If you can point me in the right direction, I'm happy to update CONTRIBUTING.md.

**Comment by chappjc** [comment]
After you work on your branch and commit, it may not be building on remote master if it has changed.  And that may be OK unless there are conflicts.  Where things differ from your intuition is instead of merging in changes from master to your branch, you rebase your branch on master.  Basically you'll need a local master branch, check that out and pull lastest into that local master.  Then check out your branch again.  `git rebase master`.  Just like with merging, there may be conflicts to resolve.  You resolve one commit at a time (edit, git add [files], git rebase --continue, repeat for other commits that have conflicts).  Then force push.

**Comment by buck54321** [comment]
Actually, I think this resolves #746 as well.

File: db/dcrsqlite/sync.go [hunk 1]
```diff
@@ -405,3 +409,54 @@ func (db *wiredDB) getBlock(ind int64) (*dcrutil.Block, *chainhash.Hash, error)
 
 	return block, blockhash, nil
 }
+
+// Import all side chains. Similar to pgblockchain.MissingSideChainBlocks
```
Comment (chappjc) [inline db/dcrsqlite/sync.go]: Please begin all function docs, especially for exported functions, with the function name (i.e. `// ImportSideChains ...`).  This is just convention, not my silly rule.  :)
https://blog.golang.org/godoc-documenting-go-code
https://golang.org/doc/effective_go.html#commentary

File: main.go [hunk 2]
```diff
@@ -630,15 +633,26 @@ func _main(ctx context.Context) error {
 		return nil
 	}
 
+	log.Info("Mainchain sync complete.")
+
+	// Ensure all side chains known by dcrd are also present in the base DB
+	// and import them if they are not already there.
+	if cfg.ImportSideChains {
+		log.Info("Primary DB -> Now retrieving side chain blocks from dcrd...")
+		err := baseDB.ImportSideChains(collector)
```
Comment (chappjc) [inline main.go]: It's a minor issue in most cases, but the approach of doing sqlite first then possibly postgres is inefficient. That is,

```go
if cfg.ImportSideChains {
	// retrieve all side chains and import into sqlite
}
if usePG && cfg.ImportSideChains {
	// retrieve all side chains and import into sqlite
}
```

vs.

```go
if cfg.ImportSideChains {
	// retrieve all side chains

	// import into sqlite

	if usePG {
		// import into postgres
	}
}
```

Changing to the latter requires some refactoring, so it's not a big deal right now.  Just take note.

File: db/dcrsqlite/sqlite.go [hunk 3]
```diff
@@ -912,6 +1076,73 @@ func (db *DB) RetrieveStakeInfoExtended(ind int64) (*apitypes.StakeInfoExtended,
...
+	// Each tuple is a row with columns [cid, name, type, notnull, dflt_value, pk]
+	rows, err := db.Query(db.getBlockSummaryTableInfo)
+	if err != nil {
+		log.Errorf("Query failed: %v", err)
+		return err
+	}
+	defer rows.Close()
+	rowCounter := 0
+	for rows.Next() {
+		rowCounter += 1
```
[the remark this answers is not in the document]
Comment (buck54321) [reply db/dcrsqlite/sqlite.go]: The way I implemented this, which is probably wrong, is to use a SQL query 

`d.getBlockSummaryTableInfo = fmt.Sprintf(`PRAGMA table_info(%s);`, TableNameSummaries)`

which returns column information as rows. That way, if you wanted a stronger check, you could check to see which column has the primary key or check the column order. I'm happy to either implement the PK check or to find a different way of counting columns.

File: main.go (hunk 2 above)
Comment (buck54321) [reply main.go]: This was part of the reason I tried to implement this as a single function call in main.go. Is there any reason both of these sidechain import routines can't be implemented in asynchronously?

Comment (chappjc) [reply main.go]: It's not an issue to do it separately for sqlite and pg, but the retrieval gets run twice.  Usually this is just a handful of blocks, so no worries.  On testnet3 there was a crazy 4k block reorg early on, and I my node still remembers it -- which is great for testing these things.  But this is certainly not something that will be an issue on mainnet.

File: db/dcrsqlite/sqlite.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply db/dcrsqlite/sqlite.go]: Oh, I see!  I just read the query incorrectly.  The pragma approach is probably perfect.  I suppose that's standard sqlite, so that's great.

**Comment by chappjc** [comment]
Another thing that is entirely my fault for not including in CONTRIBUTING.md: when there is a review and edits in response, it's helpful to just add new commits instead of squashing down. Github is good about showing changes since a review, but only if the git history remains.

Then when all is well and the PR can be merged, it's cool to squash the branch.  However, it's really OK to leave it since there is a "Squash and merge" option that I usually use.  Just in this case it's helpful to have the PR author comment here with a message for the squashed commit so I can put that message in.

File: main.go (hunk 2 above)
Comment (buck54321) [reply main.go]: K. Maybe as a separate issue, but I'm happy to do this. Probably have a wrapper function that pulls the blocks once and feeds them through a checker method of each DB object. Can that be run as a goroutine?

**Comment by buck54321** [comment]
I'll keep my CONTRIBUTING.md branch alive and update as I learn the system here. 

Do I have to do anything else once I've pushed my changes, or do you get a notification of some kind?

Also, I was a little confused on the relation of this issue to #601. I may not understand the reorg correctly, but doesn't the PGSQL sidechain check (and that amazing new SQLite check someone just did) already resolve that issue on startup?

File: api/apiroutes.go [hunk 4]
```diff
@@ -727,15 +727,21 @@ func (c *appContext) getBlockFeeInfo(w http.ResponseWriter, r *http.Request) {
 }
 
 func (c *appContext) getBlockStakeInfoExtended(w http.ResponseWriter, r *http.Request) {
-	idx := c.getBlockHeightCtx(r)
-	if idx < 0 {
+	hash := c.getBlockHashCtx(r)
```
Comment (chappjc) [inline api/apiroutes.go]: In (`NewAPIRouter`) api/apirouter.go, note that `getBlockStakeInfoExtended` is bound to several URL paths:

    rd.Get("/pos", app.getBlockStakeInfoExtended)

Depending on the full path, different middleware is used to embed the height/hash.

- Under /block/best, it is using the `BlockIndexLatestCtx` middleware, which embeds only the best block index in the request context (no hash).
- Under /block/hash/{blockhash}, it is using `BlockHashPathAndIndexCtx`, which embeds the hash from the URL path, looks up the corresponding height, and embeds that too.
- Under /block/{idx}, it is using `BlockIndexPathCtx`, which embeds the index from the path only.
- Under /block/range/{idx0}/{idx}, it is using `m.BlockIndex0PathCtx` and `m.BlockIndexPathCtx`, which embed both indexes from the path only.  Although this path prefix doesn't bind to `getBlockStakeInfoExtended`.

Basically, the `(*appContext).getWhateverCtx(*http.Request)` type functions try to pull out a value.  The middleware set for different paths decides what/when/how the values get put into those context values.

**Comment by chappjc** [comment]
OK, thanks for updating CONTRIBUTING.md as you go.

The startup reorg check is not resolved.  These changes won't import the side chain that was mainchain because it's already in the DB.  But doing so would not be correct because the mainchain needs to be disconnected carefully as in TipToSide chain so as to keep the other DBs (stakedb and a ticket database) correct.

File: main.go (hunk 2 above)
Comment (chappjc) [reply main.go]: I don't see why they can't be run concurrently.

File: api/apiroutes.go (hunk 4 above)
Comment (buck54321) [reply api/apiroutes.go]: OK. This is still a little confusing, but I think I've got it. There are now two distinct methods for the different contexts, and I think I've assigned them correctly. I did see one weird error

`2018-10-29 12:51:32.745 [ERR] RPCC: GetBlockVerbose(0000000000000000000000000000000000000000000000000000000000000000) failed: -5: Block not found: 0000000000000000000000000000000000000000000000000000000000000000`

but I wasn't able to reproduce it. I noticed that in `apirouter`, some index-only-context routes are calling `apiroutes.getBlockVerbose` which uses `getBlockHashCtx`, so I think that this might be another instance of the same error I made.

Comment (chappjc) [reply api/apiroutes.go]: The genesis (block 0) has no previous block, so it is set to that long string, `0000000000000000000000000000000000000000000000000000000000000000`
https://explorer.dcrdata.org/block/0

File: db/dcrsqlite/sync.go [hunk 5]
```diff
@@ -405,3 +409,54 @@ func (db *wiredDB) getBlock(ind int64) (*dcrutil.Block, *chainhash.Hash, error)
...
 }
+
+// ImportSideChains imports all side chains. Similar to pgblockchain.MissingSideChainBlocks
+// plus the rest from main.go
+func (db *wiredDB) ImportSideChains(collector *blockdata.Collector) error {
+	tips, err := rpcutils.SideChains(db.client)
+	if err != nil {
+		return err
+	}
+	hashlist := make([]*chainhash.Hash, 0)
```
Comment (chappjc) [inline db/dcrsqlite/sync.go]: When not allocating the slice, we generally do just `var hashlist []*chainhash.Hash`.

Comment (chappjc) [reply db/dcrsqlite/sync.go]: However, we know the length of `tips`, so we should actually do `hashlist := make([]*chainhash.Hash, 0, len(tips))` to have the right amount of space allocated.

File: db/dcrsqlite/sync.go [hunk 6]
```diff
@@ -405,3 +409,54 @@ func (db *wiredDB) getBlock(ind int64) (*dcrutil.Block, *chainhash.Hash, error)
...
+	for it := range tips {
+		sideHeight := tips[it].Height
+		log.Tracef("Primary DB -> Getting base DB side chain with tip %s at %d.", tips[it].Hash, sideHeight)
+		sideChain, err := rpcutils.SideChainFull(db.client, tips[it].Hash)
+		if err != nil {
+			log.Errorf("Primary DB -> Unable to get side chain blocks for chain tip %s: %v", tips[it].Hash, err)
+			return err
+		}
+		sideHeight -= int64(len(sideChain)) - 1
```
Comment (chappjc) [inline db/dcrsqlite/sync.go]: The computed value does not appear to be used.

Comment (chappjc) [reply db/dcrsqlite/sync.go]: Seems like this might be copy pasta, but I want to be sure a detail wasn't left out here.

**Comment by chappjc** [comment]
Code changes approved.  Just needs testing on my end.

File: db/dcrsqlite/sqlite.go [hunk 7]
```diff
@@ -912,6 +1037,99 @@ func (db *DB) RetrieveStakeInfoExtended(ind int64) (*apitypes.StakeInfoExtended,
...
+	}
+
+	// It simply checks whether there are enough columns for the current structure
+	if rowCounter >= 12 {
+		return nil
+	}
+
+	log.Info("Detected old SQLite table structure. Updating.")
```
Comment (chappjc) [inline db/dcrsqlite/sqlite.go]: It occurred to me that we should really backup the sqlt.db file first.

Comment (buck54321) [reply db/dcrsqlite/sqlite.go]: Should I delete the backup if the swap is successful? All table mods are wrapped in a transaction right now, so a failed table copy should be rolled back.

Comment (chappjc) [reply db/dcrsqlite/sqlite.go]: It's not a corrupted db I'm concerned with (as you said, failures are rolled back), but devs like us losing our DBs while testing and such.  Although there may be a need to revert to an old version of dcrdata which won't support the new DB format. So I would suggest keeping the old DB copy around regardless of the the migration result.

**Comment by buck54321** [comment]
Though there was some uncertainty regarding `os.Read` on the SQLite file, I tested on both Windows and Linux with no apparent issues.

File: db/dcrsqlite/sqlite.go [hunk 8]
```diff
@@ -929,6 +1074,120 @@ func (db *DB) RetrieveStakeInfoExtended(ind int64) (*apitypes.StakeInfoExtended,
...
+	}
+
+	err = transaction.Commit()
+	if err != nil {
+		log.Error("Failed to commit SQL transaction after table reorg.")
+		return err
+	}
+
+	// Clean up the file a little bit
+	_, err = db.Query("VACUUM;")
```
Comment (chappjc) [inline db/dcrsqlite/sqlite.go]: This needs to be `_, err = db.Exec("VACUUM;")`

Comment (chappjc) [reply db/dcrsqlite/sqlite.go]: After the table upgrade with no VACUUM, the files size was 451538944 B.
With VACUUM, the file size is reduced to 323424256 B.
